commit 9124a26401483bf2b13a99cb4317dce3f677060f
Author: Kees Cook <kees@kernel.org>
Date:   Thu Sep 29 01:58:59 2022 -0700

    kunit/fortify: Validate __alloc_size attribute results
    
    Validate the effect of the __alloc_size attribute on allocators. If the
    compiler doesn't support __builtin_dynamic_object_size(), skip the
    associated tests.
    
    (For GCC, just remove the "--make_options" line below...)
    
    $ ./tools/testing/kunit/kunit.py run --arch x86_64 \
            --kconfig_add CONFIG_FORTIFY_SOURCE=y \
            --make_options LLVM=1
            fortify
    ...
    [15:16:30] ================== fortify (10 subtests) ===================
    [15:16:30] [PASSED] known_sizes_test
    [15:16:30] [PASSED] control_flow_split_test
    [15:16:30] [PASSED] alloc_size_kmalloc_const_test
    [15:16:30] [PASSED] alloc_size_kmalloc_dynamic_test
    [15:16:30] [PASSED] alloc_size_vmalloc_const_test
    [15:16:30] [PASSED] alloc_size_vmalloc_dynamic_test
    [15:16:30] [PASSED] alloc_size_kvmalloc_const_test
    [15:16:30] [PASSED] alloc_size_kvmalloc_dynamic_test
    [15:16:30] [PASSED] alloc_size_devm_kmalloc_const_test
    [15:16:30] [PASSED] alloc_size_devm_kmalloc_dynamic_test
    [15:16:30] ===================== [PASSED] fortify =====================
    [15:16:30] ============================================================
    [15:16:30] Testing complete. Ran 10 tests: passed: 10
    [15:16:31] Elapsed time: 8.348s total, 0.002s configuring, 6.923s building, 1.075s running
    
    For earlier GCC prior to version 12, the dynamic tests will be skipped:
    
    [15:18:59] ================== fortify (10 subtests) ===================
    [15:18:59] [PASSED] known_sizes_test
    [15:18:59] [PASSED] control_flow_split_test
    [15:18:59] [PASSED] alloc_size_kmalloc_const_test
    [15:18:59] [SKIPPED] alloc_size_kmalloc_dynamic_test
    [15:18:59] [PASSED] alloc_size_vmalloc_const_test
    [15:18:59] [SKIPPED] alloc_size_vmalloc_dynamic_test
    [15:18:59] [PASSED] alloc_size_kvmalloc_const_test
    [15:18:59] [SKIPPED] alloc_size_kvmalloc_dynamic_test
    [15:18:59] [PASSED] alloc_size_devm_kmalloc_const_test
    [15:18:59] [SKIPPED] alloc_size_devm_kmalloc_dynamic_test
    [15:18:59] ===================== [PASSED] fortify =====================
    [15:18:59] ============================================================
    [15:18:59] Testing complete. Ran 10 tests: passed: 6, skipped: 4
    [15:18:59] Elapsed time: 11.965s total, 0.002s configuring, 10.540s building, 1.068s running
    
    Cc: David Gow <davidgow@google.com>
    Cc: linux-hardening@vger.kernel.org
    Signed-off-by: Kees Cook <keescook@chromium.org>

diff --git a/lib/Makefile b/lib/Makefile
index 322178b9f7fb..2f0454b931dc 100644
--- a/lib/Makefile
+++ b/lib/Makefile
@@ -378,6 +378,7 @@ CFLAGS_overflow_kunit.o = $(call cc-disable-warning, tautological-constant-out-o
 obj-$(CONFIG_OVERFLOW_KUNIT_TEST) += overflow_kunit.o
 CFLAGS_stackinit_kunit.o += $(call cc-disable-warning, switch-unreachable)
 obj-$(CONFIG_STACKINIT_KUNIT_TEST) += stackinit_kunit.o
+CFLAGS_fortify_kunit.o += $(call cc-disable-warning, unsequenced)
 obj-$(CONFIG_FORTIFY_KUNIT_TEST) += fortify_kunit.o
 obj-$(CONFIG_STRSCPY_KUNIT_TEST) += strscpy_kunit.o
 obj-$(CONFIG_SIPHASH_KUNIT_TEST) += siphash_kunit.o